### ISSUE TITLE
Thesis support

### ISSUE THREAD

**Issue by ingrinder** [typst/hayagriva#86]
At least in the IEEE format, it is customary to denote a thesis by including e.g. "Ph.D. Dissertation" or "M.D. Thesis" in the citation. Currently, Hayagriva has a `thesis` type, however, it ignores the `organization` and `location` field and the closest way to correctly reference a thesis with it is to write e.g. `publisher: Ph.D. Thesis, School of Engineering, Lancaster University, Lancaster` which is not ideal.

The intuitive way I imagine this should work would be to have `type: thesis` use the `organization` and `location` fields, then have another field for the thesis type. As an idea, perhaps the `type` field could be split up, to make something like the following, since it is kind of part of the type of the reference?

```
My-Thesis:
  type:
    value: thesis
    degree: "Ph.D. Thesis"
  title: My outstanding thesis
  author:
    - Bromilow, Michael
  date: 2023-11
  organization: School of Engineering, Lancaster University
  location: Lancaster
```

To give (in IEEE style):

`M. Bromilow, "My outstanding thesis", Ph.D. Thesis, School of Engineering, Lancaster University, Lancaster, Nov. 2023.`

**Comment by twrightsman** [comment]
CSL uses the [genre](https://docs.citationstyles.org/en/stable/specification.html#standard-variables) field to specify the type of thesis. Currently, it doesn't seem to be rendered properly in Typst 0.10.0:

```yaml
madje-2022-typst:
  type: thesis
  genre: Doctoral dissertation
  title: A Programmable Markup Language for Typesetting
  author: Mädje, Laurenz
  date: 2022-09-08
  publisher: Technical University of Berlin
  url: https://www.user.tu-berlin.de/laurmaedje/programmable-markup-language-for-typesetting.pdf
```

renders to

Mädje, L. (2022). _A Programmable Markup Language for Typesetting_ [Technical University of Berlin]. https://www.user.tu-berlin.de/laurmaedje/programmable-markup-language-for-typesetting.pdf

in APA style, which is [missing](https://apastyle.apa.org/style-grammar-guidelines/references/examples/published-dissertation-references) the "Doctoral dissertation" in the brackets.

**Comment by ingrinder** [comment]
Is `genre` valid in Hayagriva YAML files? It isn't documented, but would seem sensible to add (particularly if it's already defined in CSL).

I noticed that #112 moved from the Typst repo is a very similar issue regarding fields not being shown in the `thesis` type, perhaps these issues should be merged.

**Comment by twrightsman** [comment]
> Is `genre` valid in Hayagriva YAML files?

Ah, good point. I had assumed the Hayagriva YAML format tried to map to CSL-defined variables but I now see that's not really the case; it really is its own format. I hope Hayagriva doesn't stray too far from CSL, but that's best discussed in another issue I think.

**Comment by EpicEricEE** [comment]
The field `genre` is valid starting from hayagriva v0.5.2. To print the school, use the `publisher` field, which (starting from v0.6.0) can also take a location:
```yml
my-thesis:
  type: thesis
  genre: Ph.D. Thesis
  title: My outstanding thesis
  author: Bromilow, Michael
  date: 2023-11
  publisher:
    name: School of Engineering, Lancaster University
    location: Lancaster
```

Renders to (with IEEE style):
> M. Bromilow, "My outstanding thesis," Ph.D. Thesis, School of Engineering, Lancaster University, Lancaster, 2023.

**Comment by reknih** [comment]
Closed in #214
